backend/src/core/service/subscription.py

```diff
@@ -0,0 +1,200 @@
+
+        subscription_id = transaction.reference_id.split("_")[1]
+        subscription = await self.subscription_opr.get_subscription_by_id(subscription_id)
+        
+        if not subscription:
+            return
+        
+        # Delete user instance and subscription
+        await self.subscription_opr.delete_subscription(subscription.subscription_id)
+        await instance_service.delete_instance(instance_id=subscription.instance_id)
+    
+    async def process_overdue_subscriptions(self, overdue_subscriptions: List[Transaction]) -> None:
+        """Attempt to process all overdue subscription payments."""
+        for transaction in overdue_subscriptions:
+            await self.process_transaction(transaction)
+        
+    async def process_expired_subscriptions(self, expired_subscriptions: List[Transaction]) -> None:
+        """Mark expired subscriptions and apply penalties."""
+        for transaction in expired_subscriptions:
+            transaction.transaction_status = TransactionStatus.EXPIRED
+            transaction.last_updated_at = DateTimeUtils.now_dt()
+            await self.transaction_opr.upsert_transaction(transaction)
+            await self.apply_penalty(transaction)
+
+    # Private helper methods
+    def _calculate_payment_amount(self, hourly_cost: float) -> float:
+        """Calculate the payment amount for a subscription period."""
+        hours_in_payment_interval = PAYMENT_INTERVAL.total_seconds() / 3600  
+        return hourly_cost * hours_in_payment_interval
+
+    def _create_subscription_transaction(self, user_id: uuid.UUID, subscription_id: uuid.UUID, amount: float) -> Transaction:
+        """Create a subscription payment transaction."""
+        current_time = DateTimeUtils.now_dt()
+        return Transaction(
+            user_id=user_id,
+            transaction_type=TransactionType.SUBSCRIPTION_PAYMENT,
+            transaction_status=TransactionStatus.SCHEDULED,
+            reference_id=f"subscription_{subscription_id}",
+            amount=amount,
+            created_at=current_time,
+            last_updated_at=current_time
+        )
+
+    async def _get_subscription_transactions(
+        self, 
+        subscriptions: List[UserSubscription], 
+        transaction_statuses: List[TransactionStatus]
+    ) -> List[Transaction]:
+        """Get transactions for a list of subscriptions with specific statuses."""
+        if not subscriptions:
+            return []
+            
+        reference_ids = [f"subscription_{subscription.subscription_id}" for subscription in subscriptions]
+        
+        return await self.transaction_opr.get_transactions_by_reference_ids(
+            reference_ids=reference_ids,
+            transaction_type=[TransactionType.SUBSCRIPTION_PAYMENT],
+            transaction_status=transaction_statuses
+        )
+
+    async def _process_topup_transaction(self, transaction: Transaction, user_wallet: UserWallet) -> None:
```

Comment:
Yeah i definitely think we shouldn't do update balance like this, it just a race condition waiting to happen